// ibm-cloud-sdk-core/auth/utils/helpers.js
"use strict";
/**
 * (C) Copyright IBM Corp. 2019, 2022.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.atMostOne = exports.atLeastOne = exports.removeSuffix = exports.onlyOne = exports.getCurrentTime = exports.validateInput = exports.checkCredentials = exports.computeBasicAuthHeader = void 0;
var helper_1 = require("../../lib/helper");
/**
 * Compute and return a Basic Authorization header from a username and password.
 *
 * @param username - The username or client id
 * @param password - The password or client secret
 * @returns a Basic Auth header with format "Basic <encoded-credentials>"
 */
function computeBasicAuthHeader(username, password) {
    var encodedCreds = Buffer.from("".concat(username, ":").concat(password)).toString('base64');
    return "Basic ".concat(encodedCreds);
}
exports.computeBasicAuthHeader = computeBasicAuthHeader;
// returns true if the string has a curly bracket or quote as the first or last character
// these are common user-issues that we should handle before they get a network error
function badCharAtAnEnd(value) {
    return (value.startsWith('{') || value.startsWith('"') || value.endsWith('}') || value.endsWith('"'));
}
/**
 * Checks credentials for common user mistakes of copying \{, \}, or \" characters from the documentation
 *
 * @param obj - the options object holding credentials
 * @param credsToCheck - an array containing the keys of the credentials to check for problems
 * @returns a string with the error message if there were problems, null if not
 */
function checkCredentials(obj, credsToCheck) {
    var errorMessage = '';
    credsToCheck.forEach(function (cred) {
        if (obj[cred] && badCharAtAnEnd(obj[cred])) {
            errorMessage += "The ".concat(cred, " shouldn't start or end with curly brackets or quotes. Be sure to remove any {, }, or \"");
        }
    });
    if (errorMessage.length) {
        errorMessage +=
            'Revise these credentials - they should not start or end with curly brackets or quotes.';
        return new Error(errorMessage);
    }
    return null;
}
exports.checkCredentials = checkCredentials;
/**
 * Validates "options".
 * @param options - a configuration options object
 * @param requiredOptions - the list of properties that must be present in "options"
 *
 * @throws Error: "options" failed validation
 */
function validateInput(options, requiredOptions) {
    // check for required params
    var missingParamsError = (0, helper_1.getMissingParams)(options, requiredOptions);
    if (missingParamsError) {
        throw missingParamsError;
    }
    // check certain credentials for common user errors: username, password, and apikey
    // note: will only apply to certain authenticators
    var credsToCheck = ['username', 'password', 'apikey'];
    var credentialProblems = checkCredentials(options, credsToCheck);
    if (credentialProblems) {
        throw credentialProblems;
    }
}
exports.validateInput = validateInput;
/**
 * Gets the current time.
 *
 * @returns the current time in seconds.
 */
function getCurrentTime() {
    return Math.floor(Date.now() / 1000);
}
exports.getCurrentTime = getCurrentTime;
/**
 * Checks for only one of two elements being defined.
 * Returns true if a is defined and b is undefined,
 * or vice versa. Returns false if both are defined
 * or both are undefined.
 *
 * @param a - The first object
 * @param b - The second object
 * @returns true if and only if exactly one of a or b is defined
 */
function onlyOne(a, b) {
    return Boolean((a && !b) || (b && !a));
}
exports.onlyOne = onlyOne;
/**
 * Removes a given suffix if it exists.
 *
 * @param str - the base string to operate on
 * @param suffix - the suffix to remove, if present
 *
 * @returns the substring of "str" that remains after the suffix is removed
 */
function removeSuffix(str, suffix) {
    if (str.endsWith(suffix)) {
        str = str.substring(0, str.lastIndexOf(suffix));
    }
    return str;
}
exports.removeSuffix = removeSuffix;
/**
 * Checks for at least one of two elements being defined.
 *
 * @param a - the first object
 * @param b - the second object
 * @returns true if a or b is defined; false if both are undefined
 */
function atLeastOne(a, b) {
    return Boolean(a || b);
}
exports.atLeastOne = atLeastOne;
/**
 * Verifies that both properties are not specified.
 *
 * @param a - The first object
 * @param b - The second object
 *
 * @returns  false if a and b are both defined, true otherwise

 */
function atMostOne(a, b) {
    return Boolean(!(a && b));
}
exports.atMostOne = atMostOne;
